# Reducing Bias in Machine Learning Models:
### Modeling Part 2: Advanced Models (Base and Optimized)


#### Kristen Lo - BrainStation
---

### Table of Contents
- [Introduction](#intro)
- [Part 4: Advanced Models (Base and Optimized)](#clean)
    - 1.1: [Housekeeping](#house)
    - 1.2: [Random Forest](#ran)
    - 1.3: [xGBoost](#boost)
- [Conclusion](#conc)


---
### <a id = 'intro'></a> Introduction

In this notebook, we will be preparing our advanced models and then optimizing them for performance. 

There is a need to be able to predict the hospital admission rates for diabetic patients. However, using traditional machine learning models can lead to health disparities caused by biased data which can be related to demographic data (ex. race, age, income, insurance etc). These biases need to be removed prior to modelling so that bias isn't introduced into the model. Building on the work of Raza, S. who aimed to predict, diagnose, and mitigate health disparities in hospital re-admission, my aim is to replicate the study performed by Raza and create my own model that's able to screen for biases and predict admission rates for diabetics visiting the ER. 


Data was sourced from all adult Emergency Department visits from March 2014 - July 2017 from one academic and two community emergency rooms, apart from the Yale New Haven Health system. These visits resulted in either admission to their respective hospital or discharge. 

There are a total of 972 variables that we extracted per patient visit from 560,486 patient visits. 

Courtesy of:
 "Hong WS, Haimovich AD, Taylor RA (2018) Predicting hospital admission at emergency department triage using machine learning. PLoS ONE 13(7): e0201016." (https://doi.org/10.1371/journal.pone.0201016)




-----

## <a id = 'clean'></a> Part 3: Vanilla Modelling (Basic and Optimized) 

---
#### <a id = 'housekeeping'></a> 1.1 HouseKeeping 

Loading the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.model_selection import train_test_split,RandomizedSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score, roc_curve, f1_score
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from imblearn.over_sampling import RandomOverSampler, SMOTE, ADASYN
from sklearn.feature_selection import SelectFromModel
from imblearn.pipeline import Pipeline as ImbPipeline
from scipy.stats import uniform, randint

Loading the clean csv file

In [2]:
health_data = pd.read_csv('Data/clean_health_data_pt2.csv')

In [ ]:
pd.set_option('display.max_columns', None)  # Show all columns
pd.set_option('display.max_rows', None) #Show all rows

In [ ]:
health_data.head()

demo_gender  demo_race  disposition  pmh_2ndarymalig  pmh_abdomhernia  \
0            0          0            0                0                0   
1            0          0            1                0                0   
2            0          0            0                0                0   
3            1          0            1                0                0   
4            1          0            1                0                0   

   pmh_acqfootdef  pmh_acrenlfail  pmh_acutecvd  pmh_acutemi  pmh_acutphanm  \
0               0               0             0            0              0   
1               0               0             0            0              0   
2               0               0             0            0              0   
3               0               0             0            0              0   
4               0               0             0            0              0   

   pmh_adltrespfl  pmh_alcoholrelateddisorders  pmh_amniosdx  pmh_anemia  \
0               0                            0             0           1   
1               0                            0             0           1   
2               0                            0             0           1   
3               0                            0             0           0   
4               0                            0             0           0   

   pmh_aneurysm  pmh_anxietydisorders  pmh_artembolism  pmh_asppneumon  \
0             0                     0                0               0   
1             0                     0                0               0   
2             0                     0                0               0   
3             0                     0                0               0   
4             0                     0                0               0   

   pmh_asthma  pmh_attentiondeficitconductdisruptivebeha  pmh_biliarydx  \
0           0                                          0              0   
1           0                                          0              0   
2           0                                          0              0   
3           0                                          0              0   
4           0                                          0              0   

   pmh_bladdercncr  pmh_blindness  pmh_bnignutneo  pmh_bonectcncr  pmh_bph  \
0                0              0               0               0        0   
1                0              0               0               0        0   
2                0              0               0               0        0   
3                0              0               0               0        0   
4                0              0               0               0        0   

   pmh_brainnscan  pmh_breastcancr  pmh_breastdx  pmh_brnchlngca  \
0               0                0             0               0   
1               0                0             0               0   
2               0                0             0               0   
3               0                0             0               0   
4               0                0             0               0   

   pmh_cardiaarrst  pmh_carditis  pmh_cataract  pmh_chfnonhp  \
0                0             0             1             0   
1                0             0             1             0   
2                0             0             1             0   
3                0             0             0             0   
4                0             0             0             0   

   pmh_chrkidneydisease  pmh_coaghemrdx  pmh_coloncancer  pmh_comabrndmg  \
0                     0               0                0               0   
1                     0               0                0               0   
2                     0               0                0               0   
3                     0               0                0               0   
4                     0               0                0               0   

In [ ]:
health_data.shape

(105627, 443)

In [ ]:
health_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105627 entries, 0 to 105626
Columns: 443 entries, demo_gender to dum_huse__3-plus_hormones
dtypes: int64(443)
memory usage: 357.0 MB


Considering that all of the information is binary, int64 dtype is bloating the file unnecessarily. Let's convert it to int8 to conserve memory and assist in modeling.

In [ ]:
# Assuming df is your DataFrame
int64_columns = health_data.select_dtypes(include='int64').columns

# Convert selected columns to int8
health_data[int64_columns] = health_data[int64_columns].astype('int8')

In [ ]:
health_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105627 entries, 0 to 105626
Columns: 443 entries, demo_gender to dum_huse__3-plus_hormones
dtypes: int8(443)
memory usage: 44.6 MB


As we can see, the memory usage has decreased from 357 mb to 44.6mb. This should be less computationally intensive.

---
#### <a id = 'log'></a> 1.2 Logistic Regression 